File: backend/app/schemas/calendars.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

# --- Calendar ---

CalendarGenerationMode = Literal["manual", "assisted_auto", "scheduled_auto"]


class CalendarRead(BaseModel):
    id: str
    year: int
    month: int
    status: str
    generation_mode: CalendarGenerationMode
    created_by: str | None
    approved_by: str | None
    created_at: datetime
    updated_at: datetime
    approved_at: datetime | None

    model_config = {"from_attributes": True}


class CreateCalendarRequest(BaseModel):
    year: int = Field(ge=2020, le=2100)
    month: int = Field(ge=1, le=12)
    generation_mode: CalendarGenerationMode = "manual"


# --- Calendar Version ---

class CalendarVersionRead(BaseModel):
    id: str
    calendar_id: str
    version_number: int
    status: str
    created_by: str | None
    reason: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Calendar Assignment ---

class CalendarAssignmentRead(BaseModel):
    id: str
    calendar_version_id: str
    service_date: date
    service_start_at: datetime | None
    service_area_id: str
    doctor_id: str
    assignment_source: str
    rationale: dict | None
    override_justification: str | None
    created_by: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AssignDoctorRequest(BaseModel):
    service_date: date
    service_area_id: str
    doctor_id: str
    override_justification: str | None = Field(default=None, max_length=500)
    force_warnings: list[str] | None = Field(default=None)


class ReplaceAssignmentRequest(BaseModel):
    doctor_id: str
    override_justification: str | None = Field(default=None, max_length=500)
    force_warnings: list[str] | None = Field(default=None)


# --- Eligible Doctors ---

class EligibleDoctorRead(BaseModel):
    id: str
    full_name: str
    specialty: str | None = None
    rank_name: str | None = None
    altera_orden: bool | None = None  # None = sin patrón, False = respeta, True = altera

    model_config = {"from_attributes": True}


class EligibleDoctorsResponse(BaseModel):
    doctors: list[EligibleDoctorRead]


# --- Slot Evaluation ---

class HardBlockItem(BaseModel):
    code: str
    description: str


class WarningItem(BaseModel):
    code: str
    description: str


class EvaluationRequest(BaseModel):
    doctor_id: str
    service_date: date
    service_area_id: str


class EvaluationResponse(BaseModel):
    hard_blocks: list[HardBlockItem]
    warnings: list[WarningItem]


# --- Calendar View (grid for UI) ---

class DaySlot(BaseModel):
    service_date: date
    service_area_id: str
    assignment: CalendarAssignmentRead | None
    has_warning: bool = False
    warning_message: str | None = None


class CalendarGridResponse(BaseModel):
    calendar: CalendarRead
    version: CalendarVersionRead
    slots: list[DaySlot]
    gaps: list[dict]


# --- Approval ---

class ApproveCalendarRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=500)


class UnresolvedGapRead(BaseModel):
    id: str
    calendar_version_id: str
    service_date: date
    service_area_id: str
    reason_code: str
    description: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Calendar Week ---

class DoctorAssignmentCountRead(BaseModel):
    doctor_id: str
    doctor_name: str
    count: int


class WeekRead(BaseModel):
    id: str
    week_number: int
    label: str
    start_date: str
    end_date: str
    status: str
    assignment_count: int = 0
    doctor_assignment_counts: list[DoctorAssignmentCountRead] = Field(default_factory=list)
    approved_by: str | None = None
    approved_at: str | None = None

    model_config = {"from_attributes": True}


class ApproveWeekRequest(BaseModel):
    notes: str | None = None


# --- Equilibrio / Pattern tracking ---


class EquilibrioItem(BaseModel):
    doctor_id: str
    nombre_medico: str
    servicios_asignados: int
    servicios_esperados: int
    alteraciones_al_orden: int


class EquilibrioResponse(BaseModel):
    mes: int
    ano: int
    items: list[EquilibrioItem]

File: frontend/src/api/calendars.ts
import { apiFetch } from "./client";

export interface GenerationSlotResult {
  service_date: string;
  service_area_id: string;
  assigned_doctor_id: string | null;
  warnings: string[];
  score: number | null;
}

export interface GenerationResponse {
  version_id: string;
  calendar_id: string;
  month: number;
  year: number;
  calendar_status: "draft" | "partial" | "approved";
  generation_mode: CalendarRead["generation_mode"];
  review_required: boolean;
  total_slots: number;
  assigned_count: number;
  gap_count: number;
  slots: GenerationSlotResult[];
}

export interface CalendarRead {
  id: string;
  year: number;
  month: number;
  status: "draft" | "partial" | "approved";
  generation_mode: "manual" | "assisted_auto" | "scheduled_auto";
  created_by: string | null;
  approved_by: string | null;
  created_at: string;
  updated_at: string;
  approved_at: string | null;
}

export interface CalendarVersionRead {
  id: string;
  calendar_id: string;
  version_number: number;
  status: "draft" | "approved";
  created_by: string | null;
  reason: string | null;
  created_at: string;
}

export interface CalendarAssignmentRead {
  id: string;
  calendar_version_id: string;
  service_date: string;
  service_area_id: string;
  doctor_id: string;
  override_justification: string | null;
  assignment_source: string;
  rationale: Record<string, unknown> | null;
  created_by: string | null;
  created_at: string;
}

export interface DaySlot {
  service_date: string;
  service_area_id: string;
  assignment: CalendarAssignmentRead | null;
  has_warning: boolean;
  warning_message: string | null;
}

export interface CalendarGridResponse {
  calendar: CalendarRead;
  version: CalendarVersionRead;
  slots: DaySlot[];
  gaps: Record<string, unknown>[];
}

export interface WeekRead {
  id: string;
  week_number: number;
  label: string;
  start_date: string;
  end_date: string;
  status: string;
  assignment_count: number;
  doctor_assignment_counts: { doctor_id: string; doctor_name: string; count: number }[];
  approved_by: string | null;
  approved_at: string | null;
}

export interface EligibleDoctorRead {
  id: string;
  full_name: string;
  specialty: string | null;
  rank_name: string | null;
}

export interface EligibleDoctorsResponse {
  doctors: EligibleDoctorRead[];
}

export interface HardBlockItem {
  code: string;
  description: string;
}

export interface WarningItem {
  code: string;
  description: string;
}

export interface EvaluationRequest {
  doctor_id: string;
  service_date: string;
  service_area_id: string;
}

export interface EvaluationResponse {
  hard_blocks: HardBlockItem[];
  warnings: WarningItem[];
}

export const calendarsApi = {
  list: () => apiFetch<CalendarRead[]>("/calendars"),

  create: (
    year: number,
    month: number,
    generationMode: CalendarRead["generation_mode"] = "manual",
  ) =>
    apiFetch<CalendarRead>("/calendars", {
      method: "POST",
      body: JSON.stringify({ year, month, generation_mode: generationMode }),
    }),

  getGrid: (calendarId: string) =>
    apiFetch<CalendarGridResponse>(`/calendars/${calendarId}/grid`),

  approve: (calendarId: string, reason?: string) =>
    apiFetch<CalendarVersionRead>(`/calendars/${calendarId}/approve`, {
      method: "POST",
      body: JSON.stringify({ reason: reason ?? null }),
    }),

  unlock: (calendarId: string) =>
    apiFetch<CalendarVersionRead>(
      `/calendars/${calendarId}/unlock`,
      { method: "POST" },
    ),

  assignDoctor: (
    calendarId: string,
    versionId: string,
    payload: { service_date: string; service_area_id: string; doctor_id: string; override_justification?: string | null; force_warnings?: string[] | null },
  ) =>
    apiFetch<CalendarAssignmentRead>(
      `/calendars/${calendarId}/versions/${versionId}/assignments`,
      { method: "POST", body: JSON.stringify(payload) },
    ),

  removeAssignment: (calendarId: string, versionId: string, assignmentId: string) =>
    apiFetch<void>(
      `/calendars/${calendarId}/versions/${versionId}/assignments/${assignmentId}`,
      { method: "DELETE" },
    ),

  replaceAssignment: (
    calendarId: string,
    versionId: string,
    assignmentId: string,
    doctorId: string,
    overrideJustification?: string | null,
    forceWarnings?: string[] | null,
  ) =>
    apiFetch<CalendarAssignmentRead>(
      `/calendars/${calendarId}/versions/${versionId}/assignments/${assignmentId}`,
      { method: "PATCH", body: JSON.stringify({ doctor_id: doctorId, override_justification: overrideJustification ?? null, force_warnings: forceWarnings ?? null }) },
    ),

  generate: (calendarId: string) =>
    apiFetch<GenerationResponse>(`/calendars/${calendarId}/generate`, { method: "POST" }),

  delete: (calendarId: string) =>
    apiFetch<void>(`/calendars/${calendarId}`, { method: "DELETE" }),

  eligibleDoctors: (calendarId: string, date: string, areaId: string) =>
    apiFetch<EligibleDoctorsResponse>(
      `/calendars/${calendarId}/eligible-doctors?date=${date}&area_id=${areaId}`,
    ),

  evaluate: (calendarId: string, payload: EvaluationRequest) =>
    apiFetch<EvaluationResponse>(`/calendars/${calendarId}/evaluate`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  listWeeks: (calendarId: string) =>
    apiFetch<WeekRead[]>(`/calendars/${calendarId}/weeks`),

  approveWeek: (calendarId: string, weekId: string) =>
    apiFetch<WeekRead>(`/calendars/${calendarId}/weeks/${weekId}/approve`, {
      method: "POST",
    }),

  unlockWeek: (calendarId: string, weekId: string) =>
    apiFetch<WeekRead>(`/calendars/${calendarId}/weeks/${weekId}/unlock`, {
      method: "POST",
    }),

  exportWeeklyPDF: (calendarId: string, weekId: string) =>
    apiFetch<Blob>(
      `/reports/calendar/${calendarId}/weeks/${weekId}/pdf`,
      {},
      "blob",
    ),

  exportFullCalendarPDF: (calendarId: string) =>
    apiFetch<Blob>(
      `/reports/calendar/${calendarId}/full-pdf`,
      {},
      "blob",
    ),
};
